program/src/processor/stake_tokens.rs

```diff
@@ -12,21 +25,167 @@ use crate::instruction::accounts::{Context, StakeTokensAccounts};
+        ctx.accounts.validator_vote.key,
+        ctx.accounts.config.key,
+        program_id,
+    );
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    let mut stake_data = ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(&mut stake_data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    let mint_data = ctx.accounts.mint.try_borrow_data()?;
+    let mint = PodStateWithExtensions::<PodMint>::unpack(&mint_data)?;
+    let decimals = mint.base.decimals;
+
+    // Update the config and stake account data.
+    //
+    // Note: validate the amount against the total SOL staked on the validator.
+
+    require!(amount > 0, StakeError::InvalidAmount);
+
+    config.token_amount_delegated = config
+        .token_amount_delegated
+        .checked_add(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    stake.amount = stake
+        .amount
+        .checked_add(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    // Transfer the tokens to the vault (stakes them).
+
+    drop(mint_data);
+    drop(vault_data);
+
+    let mut transfer_ix = transfer_checked(
+        ctx.accounts.token_program.key,
+        ctx.accounts.source_token_account.key,
+        ctx.accounts.mint.key,
+        ctx.accounts.vault.key,
+        ctx.accounts.token_account_authority.key,
+        &[],
+        amount,
+        decimals,
+    )?;
+
+    ctx.remaining_accounts.iter().for_each(|account| {
+        transfer_ix.accounts.push(AccountMeta {
+            is_signer: account.is_signer,
+            is_writable: account.is_writable,
+            pubkey: *account.key,
+        });
+    });
+
+    let mut account_infos: Vec<AccountInfo> = Vec::with_capacity(5 + ctx.remaining_accounts.len());
+    account_infos.push(ctx.accounts.token_program.clone());
+    account_infos.push(ctx.accounts.source_token_account.clone());
+    account_infos.push(ctx.accounts.mint.clone());
+    account_infos.push(ctx.accounts.vault.clone());
+    account_infos.push(ctx.accounts.token_account_authority.clone());
+    account_infos.extend_from_slice(ctx.remaining_accounts);
+
+    invoke(&transfer_ix, &account_infos)
```

Comment:
You can use Token-2022's helper for this!
https://github.com/paladin-bladesmith/lockup-program/blob/52849fc8b0878fa76f37d6accad06723667d9268/program/src/processor.rs#L114-L124

Reply:
Nice! I was thinking that there should have been a helper for this.

Reply:
This looks way better:
```rust
spl_token_2022::onchain::invoke_transfer_checked(
   &spl_token_2022::ID,
   ctx.accounts.source_token_account.clone(),
   ctx.accounts.mint.clone(),
   ctx.accounts.vault.clone(),
   ctx.accounts.token_account_authority.clone(),
   ctx.remaining_accounts,
   amount,
   decimals,
   &[],
)
```